kv.go

```diff
@@ -10,83 +10,155 @@ import "context"
-	// Start must be lexigraphically less
-	// then end, unless `nil` is used.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically greater than or
+	// equal to this value.
 	//
-	// Start defined as nil will iterate from the first
-	// key in the store.
-	// End defined as nil will iterate till the last key
-	// in the store.
+	// Providing an `End` value equal to or smaller than this value
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
 	Start []byte
-	End   []byte
 
-	// Reverse the direction of the iteration.
+	// If Prefix is nil, and Start is provided, the iterator will
+	// only yield items with a key lexographically smaller than this
+	// value.
+	//
+	// Providing an End value equal to or smaller than Start
+	// will result in undefined behaviour:
+	// https://github.com/sourcenetwork/corekv/issues/32
+	End []byte
+
+	// Reverse the direction of the iteration, returning items in
+	// lexographically descending order of their keys.
 	Reverse bool
 
-	// Only iterate through keys. Calling Value() on the
-	// iterator *must* return nil (no error).
+	// Only iterate through keys. Calling Value on the
+	// iterator will return nil and no error.
+	//
+	// This option is currently untested:
+	// https://github.com/sourcenetwork/corekv/issues/34
+	//
+	// It is very likely ignored for the memory store iteration:
+	// https://github.com/sourcenetwork/corekv/issues/33
 	KeysOnly bool
 }
 
+// Reader contains read-only functions for interacting with a store.
 type Reader interface {
+	// Get returns the value at the given key.
+	//
+	// If no item with the given key is found, nil, and an [ErrNotFound]
+	// error will be returned.
 	Get(ctx context.Context, key []byte) ([]byte, error)
+
+	// Has returns true if an item at the given key is found, otherwise
+	// will return false.
 	Has(ctx context.Context, key []byte) (bool, error)
+
+	// Iterator returns a read-only iterator using the given options.
 	Iterator(ctx context.Context, opts IterOptions) Iterator
 }
 
+// Writer contains functions for mutating stuff stored within a store.
 type Writer interface {
+	// Set sets the value stored against the given key.
+	//
+	// If an item already exists at the given key it will be overwritten.
 	Set(ctx context.Context, key, value []byte) error
+
+	// Delete removes the value at the given key.
+	//
+	// If no matching key is found the behaviour is undefined:
+	// https://github.com/sourcenetwork/corekv/issues/36
 	Delete(ctx context.Context, key []byte) error
 }
 
+// Iterator is a read-only iterator that allows iteration over the underlying
+// store (or a part of it).
 type Iterator interface {
+	// The behaviour of Domain is undefined and varies heavily depending on
+	// which store implementation is being iterated over:
+	// https://github.com/sourcenetwork/corekv/issues/31
 	Domain() (start []byte, end []byte)
```

Comment:
discussion: I think this should be removed.  It's behaviour is unclear, and I see the return type as flawed, as it is trying to massage a `Prefix` into a `start`/`end`.

We could potentially replace it with a `Options() IterOptions` func, but I am currently in favour of not bothering with that for now and just delete it.

Reply:
I agree with this sentiment, the `Domain` method came from inspiration from another package, but its utility is a little unclear. 

I might need to double check if `corekv` is used within the SourceHub side of things does the lack of `Domain` cause problems there. Since the method comes from the cosmos-db interface (if my memory is correct).

Reply:
If it's not needed, I agree that it should be removed.